jaxpm/distributed.py

```diff
@@ -0,0 +1,212 @@
+    return ((halo_x, halo_y, (0, 0)), (halo_x_ext, halo_y_ext))
+
+
+def halo_exchange(x, halo_extents, halo_periods=(True, True)):
+    if (halo_extents[0] > 0 or halo_extents[1] > 0):
+        return jaxdecomp.halo_exchange(x, halo_extents, halo_periods)
+    else:
+        return x
+
+
+def slice_unpad_impl(x, pad_width):
+
+    halo_x, _ = pad_width[0]
+    halo_y, _ = pad_width[1]
+    # Apply corrections along x
+    x = x.at[halo_x:halo_x + halo_x // 2].add(x[:halo_x // 2])
+    x = x.at[-(halo_x + halo_x // 2):-halo_x].add(x[-halo_x // 2:])
+    # Apply corrections along y
+    x = x.at[:, halo_y:halo_y + halo_y // 2].add(x[:, :halo_y // 2])
+    x = x.at[:, -(halo_y + halo_y // 2):-halo_y].add(x[:, -halo_y // 2:])
+
+    unpad_slice = [slice(None)] * 3
+    if halo_x > 0:
+        unpad_slice[0] = slice(halo_x, -halo_x)
+    if halo_y > 0:
+        unpad_slice[1] = slice(halo_y, -halo_y)
+
+    return x[tuple(unpad_slice)]
+
+
+def slice_pad(x, pad_width, sharding):
+    gpu_mesh = sharding.mesh if sharding is not None else None
+    if not gpu_mesh is None and not (gpu_mesh.empty) and (
+            pad_width[0][0] > 0 or pad_width[1][0] > 0):
+        assert sharding is not None
+        spec = sharding.spec
+        return shard_map((partial(jnp.pad, pad_width=pad_width)),
+                         mesh=gpu_mesh,
+                         in_specs=spec,
+                         out_specs=spec)(x)
+    else:
+        return x
+
+
+def slice_unpad(x, pad_width, sharding):
+    mesh = sharding.mesh if sharding is not None else None
+    if not mesh is None and not (mesh.empty) and (pad_width[0][0] > 0
+                                                  or pad_width[1][0] > 0):
+        assert sharding is not None
+        spec = sharding.spec
+        return shard_map(partial(slice_unpad_impl, pad_width=pad_width),
+                         mesh=mesh,
+                         in_specs=spec,
+                         out_specs=spec)(x)
+    else:
+        return x
+
+
+def get_local_shape(mesh_shape, sharding=None):
+    """ Helper function to get the local size of a mesh given the global size.
+  """
+    gpu_mesh = sharding.mesh if sharding is not None else None
+    if gpu_mesh is None or gpu_mesh.empty:
+        return mesh_shape
+    else:
+        pdims = gpu_mesh.devices.shape
+        return [
+            mesh_shape[0] // pdims[0], mesh_shape[1] // pdims[1],
+            *mesh_shape[2:]
+        ]
+
+
+def zeros(mesh_shape, sharding=None):
+    gpu_mesh = sharding.mesh if sharding is not None else None
+    if not gpu_mesh is None and not (gpu_mesh.empty):
+        local_mesh_shape = get_local_shape(mesh_shape, sharding)
+        spec = sharding.spec
+        return shard_map(
+          partial(jnp.zeros, shape=(local_mesh_shape), dtype='float32'),
+          mesh=gpu_mesh,
+          in_specs=(),
+          out_specs=spec)()  # yapf: disable
+    else:
+        return jnp.zeros(mesh_shape)
+
+
+def __axis_names(spec):
```

Comment:
Double leading underscore is barely used.
For semi private a single leading underscore is more traditional.